api/agent/agent.go

```diff
@@ -737,114 +745,183 @@ func newSizerRespWriter(max uint64, rw http.ResponseWriter) http.ResponseWriter
 
-func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state ContainerState) {
+func (a *agent) waitInitialize(ctx context.Context, caller slotCaller, evictor *EvictToken, udsWait <-chan error) (bool, error) {
+	for {
+		select {
+		case err := <-udsWait:
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		case <-ctx.Done(): // container shutdown
+			return false, nil
+		case <-a.shutWg.Closer(): // agent shutdown
+			return false, nil
+		case <-caller.done: // original caller disconnected or serviced by another container?
+			evictor.SetEvictable(true)
+			caller.done = nil // block 'doneChan' after this point
+		case <-evictor.C: // eviction
+			return false, nil
+		}
+	}
+}
+
+// Try to queue an error to the error channel if possible.
+func tryQueueErr(err error, ch chan error) error {
+	if err != nil {
+		select {
+		case ch <- err:
+		default:
+		}
+	}
+	return err
+}
+
+func (a *agent) runHot(ctx context.Context, caller slotCaller, call *call, tok ResourceToken, state ContainerState) {
 	// IMPORTANT: get a context that has a child span / logger but NO timeout
 	// TODO this is a 'FollowsFrom'
 	ctx = common.BackgroundContext(ctx)
 	ctx, span := trace.StartSpan(ctx, "agent_run_hot")
 	defer span.End()
 
-	// IMPORTANT: evict token is deleted *after* resource token in defer statements below.
-	// This ordering allows resource token to be freed first, which means once evict token
-	// is deleted, eviction is considered to be completed.
+	var container *container
+	var cookie drivers.Cookie
+	var waiter drivers.WaitResult
+	var runRes drivers.RunResult
+	var err error
+
+	id := id.New().String()
+	logger := logrus.WithFields(logrus.Fields{"id": id, "app_id": call.AppID, "fn_id": call.FnID, "image": call.Image, "memory": call.Memory, "cpus": call.CPUs, "idle_timeout": call.IdleTimeout})
+	ctx, cancel := context.WithCancel(common.WithLogger(ctx, logger))
+
+	initialized := make(chan struct{}) // initialized means (docker pulled, container is created/started, uds is ready)
+	udsWait := make(chan error, 1)     // track UDS state and errors
+	errQueue := make(chan error, 1)    // errors to be reflected back to the slot queue
+
 	evictor := a.evictor.CreateEvictToken(call.slotHashId, call.Memory+uint64(call.TmpFsSize), uint64(call.CPUs))
-	defer a.evictor.DeleteEvictToken(evictor)
 
 	statsUtilization(ctx, a.resources.GetUtilization())
+	state.UpdateState(ctx, ContainerStateStart, call.slots)
+
+	// stack unwind spelled out with strict ordering below.
 	defer func() {
-		statsUtilization(ctx, a.resources.GetUtilization())
-	}()
+		// IMPORTANT: we do not emit eviction errors to slot queue. Remember evictions occur
+		// in cases of overcapacity. If an accidental eviction occurs, then the client
+		// will have to wait hot poll period to spawn a new container.
+		if !evictor.isEvicted() {
+			select {
+			case <-initialized:
+			default:
+				tryQueueErr(models.ErrContainerInitFail, errQueue)
```

Comment:
this appears to queue an eviction error to the error queue just before dequeueing it to send to the slot queue, which the notice says shouldn't happen. not sure if comment or code or my understanding should change

Reply:
Only if no eviction occurs, then we try to process errors. Adjusted the comments to make this a bit clearer.